### DISCUSSION TITLE
NPM is not defined on the PATH when running `using: "node"`

### DISCUSSION THREAD

**Discussion by tiagoboeing** [actions/runner#2037] [Q&A] [answered]
This is a question about the expected behavior when using [JavaScript Custom Actions](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action).

I have a custom Action that runs using NodeJS and needs to install another CLI tool at the runtime, like `azure-functions-core-tools`, but looks like the NPM is not defined to PATH when using NodeJS runtime.

### Custom Action using NodeJS runtime

```yml
runs:
  using: "node16"
  main: "dist/index.js"
```

### Install call using `@actions/exec`

```ts
import { exec } from '@actions/exec'

await exec('npm install --location=global azure-functions-core-tools@4')
```

Received error:

> Error: Unable to locate executable file: npm. Please verify either the file path exists or the file can be found within a directory specified by the PATH environment variable. Also check the file mode to verify the file is executable.

Is this correct for custom actions that run using NodeJS? The NodeJS should be automatically defined to PATH. (Or not?)

The workaround is to define a custom action to use `using: composite` and do a pure call to NodeJS (`node ./dist/index.js`) passing all inputs using `env` (`INPUT_ABC: abc`) and after install the Node through [`@actions/tool-cache`](https://github.com/actions/toolkit/tree/main/packages/tool-cache), but this sounds stranger because I'm using a NodeJS custom actions to install NodeJS. 🤔

**Comment by tamish-max** [comment]
What `using: node16` / `node20` gives you is a Node runtime to execute the action entrypoint. It does **not** guarantee that `npm` is available on `PATH` as a standalone executable in every environment.

So this can happen, especially on self-hosted or stripped-down runners.

Two practical options:

1. Install Node/npm in the *workflow* before the action runs (for example with `actions/setup-node`).
2. Inside the action, use `@actions/tool-cache` to download Node and prepend its `bin` directory to `PATH` before calling `npm`.

Also, with `@actions/exec`, pass the command and args separately:

```ts
await exec('npm', ['install', '--location=global', 'azure-functions-core-tools@4'])
```

If you want the action to be portable, I’d avoid doing global npm installs inside a JS action unless the action bootstraps Node/npm itself or documents `setup-node` as a prerequisite.

**Comment by gistrec** [comment, marked as answer]
This is expected for JavaScript actions

When you use:

```yaml
runs:
  using: "node16"
  main: "dist/index.js"
```

the runner invokes the Node.js binary for the action, but it does not guarantee that `npm` is available on `PATH`. JavaScript actions are generally expected to be packaged/self-contained, with their dependencies bundled into `dist/` at build time

For normal JS dependencies, the recommended approach is to install them during development/CI and bundle the action, for example with `@vercel/ncc`:

```bash
npm ci
npx @vercel/ncc build index.js -o dist
```

Then commit the generated `dist/` file.

If your action really needs to install or call an external CLI at runtime, a composite action is probably a better fit, because you can explicitly prepare the environment first:

```yaml
runs:
  using: "composite"
  steps:
    - uses: actions/setup-node@v4
      with:
        node-version: "20"

    - run: npm install --location=global azure-functions-core-tools@4
      shell: bash

    - run: node ./dist/index.js
      shell: bash
```

So in short: `node` being available for a JavaScript action does not imply that `npm` is also available on `PATH`. For JS package dependencies, bundle them; for runtime CLI installation, use a composite action or install the CLI in a separate workflow step
